### PR TITLE
PRFC: Add a new Environment kind to the catalog model

### PR DESCRIPTION
**Pull request by sennyeya**
## Hey, I just made a Pull Request!

Trying to move discussion on #16389 forward 😁 . This PR adds a new `EnvironmentEntityV1Alpha` to the catalog model, relations on the component model to point to the environment, and the ability to override annotations on the Component. Currently, those overrides only correspond to front end values, ie front end plugins like PagerDuty or Sentry that use an annotation to grab xyz content.

There's a toy example of this working in https://backstage-default-pr-21560-c2185.uclusters.app.uffizzi.com/catalog/default/component/podcast-api/errors. There is a sentry config set up for prod but not for dev, so changing the picker updates the sentry table to show errors for prod only.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/catalog-model/src/kinds/ComponentEntityV1alpha1.ts [hunk 1]
```diff
@@ -39,6 +39,21 @@ export interface ComponentEntityV1alpha1 extends Entity {
...
+    /**
+     * A list of entity refs with kind=Environment, that this component is deployed to.
+     */
+    deploysToEnvironments?: string[];
+
+    /**
+     * A map of Environment entity ref to entity metadata to override in that environment. By default,
+     *  a Component's metadata reflects their production deployment.
+     */
+    environmentOverrides?: Record<
```
Comment (sennyeya) [inline packages/catalog-model/src/kinds/ComponentEntityV1alpha1.ts:51]: I'm not 100% on this being a useful way to override information. It adds a massive amount of complexity for backend processors/providers/services but is useful for frontend switching of context like PagerDuty and Sentry. Open to ideas on improvements.

File: plugins/catalog-react/src/hooks/useEntityEnvironment.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+ */
+export const EntityEnvironmentContext = createContext<
+  EntityEnvironmentContextProps | undefined
+>(undefined);
+
+/**
+ * Provides entities and filters for a catalog listing.
+ * @public
+ */
+export const EntityEnvironmentProvider = (props: PropsWithChildren<{}>) => {
```
Comment (sennyeya) [inline plugins/catalog-react/src/hooks/useEntityEnvironment.tsx:53]: Basic idea is that you can switch between environments on the front end to see information for a specific environment. This still needs more tailoring for an entity-by-entity experience, like only showing options relevant the current entity.

File: plugins/catalog/src/components/EntityLayout/EntityLayout.tsx [hunk 3]
```diff
@@ -252,61 +254,64 @@ export const EntityLayout = (props: EntityLayoutProps) => {
...
-      </Header>
+      <EntityEnvironmentProvider>
+        <Header
+          title={<EntityLayoutTitle title={headerTitle} entity={entity!} />}
+          pageTitleOverride={headerTitle}
+          type={headerType}
+        >
+          {entity && (
+            <>
+              <EntityEnvironmentPicker />
```
Comment (sennyeya) [inline plugins/catalog/src/components/EntityLayout/EntityLayout.tsx:265]: Provide a way for users to update the environment in the header. The UI will need some iteration I think 😁

**Comment by jcarres-mdsol** [comment]
With this PR, we will be able to know that "UserService" is deployed to "production" for instance.
Which is good but if the goal is to know what environment(s) something is deployed to, we could just use an annotation and call it a day.

I think the very least we need a ComponentEnvironment or such thing. 
I want to get a "ProductionUserService" entity to which I can attach things via annotations or other ways and which I can display independently. For instance pagerduty or links to dashboards would go here. Anything that smells ops would go here.
These links would be different for "StagingUserService" or other ComponentEnvironments.

**Comment by sennyeya** [comment]
@jcarres-mdsol You're early to the party 😁 , was planning to open this up on Monday. I'm expecting this PR to be 🌶️ , thanks for starting us out! 

My goal with this PR was to implement what Grendel had recommended in the attached thread which seemed to have community alignment. Most of the conversation there was about creating the initial foundation for tracking the environment side of deployments or SDLC lifecycle, `deployedTo` and `hasDeployment`. The next step here would be aggregate dashboards for these things on both the component and environment. Yes, these "relations" could be annotations, but IMO we lose a lot of richness from that. An example that comes to mind is the next layer of modeling, like what Resources/Components/(Templates?) makes up an Environment?

As you call out, I've tied the definition of Environment quite closely to Component. I think this makes sense to do now, but with an expectation that Resources could also be tied into this (and to a lesser degree Systems). From a modeling perspective, having separate kinds for each different `{kind}Environment` seems a little overkill, especially as the Environment entity itself is pretty light already.

One of the things that I was trying to figure out while writing this PR is with per-environment overrides, what does `metadata.annotations` (or any other overrideable field) refer to? Is it production, dev, some inbetween? Do we want to make this more explicit? Or does that explode the complexity of the data we store in the catalog? Ideally, any solution here would require updates to existing catalog-info files or any massive Component shape updates.

**Comment by jcarres-mdsol** [comment]
To me a "production" entity only is not particularly useful. We have a lot of stuff in production and listing them together is nor particularly interesting.

If you think of an Entity page for Production what can you think of? Maybe some big numbers "200 services" "100 webpages" or something like this, which fun but not useful.

To me what's interesting is modeling a component that exist in this or that environment.
Example:

UserServiceProduction
- Who is on call and service status
- Link to dashboards
- Link to retrieve logs, traces, metrics
- Last deploy, version deployed, link to code commit/tag
- Cost, number of deployments, customer incidents or whatever other stat per week

I can imagine in my development/sandbox Entity page. I will have similar widgets:
- Link to sandbox dashboards
- Deployments, branches, etc.
- Stats of sandbox usage, cost, etc.

Then it is ok these connect with the component on one side and a generic concept of environment on the other

**Comment by freben** [comment]
The environment catalog page could display personal and anomalous slices of the full data.

"Here are the three things you own and their real time statuses / metrics in this particular env"

"Here is a currently ongoing incident that affects this env, possibly explaining an anomaly you're seeing in your own stuff"

Etc

**Comment by freben** [comment]
But you are right that indeed, just knowing that an entity is represented in a given set of envs feels like a very minor part of the puzzle, and doesn't solve enough to be really useful. This is reflected in the issue discussion. We already get all of the above info in our instance, despite not explicitly having the envs represented in the catalog.

What seems to be the bigger need, is to represent _aspects of your entity info_ that actually _are different_ across envs. Like, having a `metadata.links` entry to the appropriate monitoring page, that's different for each env, or defining different lifecycle value for each. That, to me, starts to sound like an entirely different, larger, problem space that needs even more foundational changes.

I _think_ that's what @its02003 alludes to in https://github.com/backstage/backstage/issues/16389#issuecomment-1819928947, that maybe we'd have to invent some form of per-env overrides mechanism to actually reap value out of all of this. Not sure what the implementation would be for anything like that, but it might be in the form of the complex environment spec that that comment mentions, or in the form of allowing metadata _on relations themselves_ which is an idea that has been floated a couple of times during the years.

I'm still, I think, vaguely feeling that _only_ adding an Environment kind and relations to them, only takes us a fraction of the way, and if it proves that the full solution needs something entirely dfferent in the end, then I hope the original effort didn't go to waste.

**Comment by sennyeya** [comment]
Thanks for the early discussion all! I'll be marking this as ready for review and tagging a few folks from the original discussion (@Grendel, @regicsolutions)

**Comment by sennyeya** [comment]
@freben Curious your thoughts on the new [`spec.environmentOverrides`](https://github.com/backstage/backstage/pull/21560/files#diff-db781b59257a6c63f51e4cc5accfd3c8053f3ee08f260ad5fbb3fe8772d40972R51) added in this PR? I agree only adding the kind and relations does little to actually present a usable foundation for future work.

**Comment by Grendel** [comment]
Hey guys. Great to see the progress on this... my 2d on the discussion so far. I am still of the mind that just having an Environment kind and the ability to cross link this to components _is_ still very useful, and a great foundation to build from. Domain and System don't offer more than this kind of semantic grouping/navigation context.

I think there are some generic resources/links etc that can be collected against this. It also provides a view port/section to start building out environment specific versions of some of the existing plugins like CI/CD and cost aggregation/visualisation at environment level. 

> allowing metadata _on relations themselves_ which is an idea that has been floated a couple of times during the years

Capturing environment specific component metadata on relations seems like an elegant extension to me. "Last commit hash", "last deployment date/time", "last deployed by", "version" type attributes when we create/update the deployment relation for example?  

I would probably be looking to provide some kind of page under environment that provides any kind of component environment specific context. e.g. a series of summary panels, one per service, with key support details, environment specific URLs, last deployment details, etc...

Hope this helps.

File: packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema",
+  "$id": "EnvironmentV1alpha1",
+  "description": "An Environment represents a single static location that software is deployed to",
+  "examples": [
+    {
+      "apiVersion": "backstage.io/v1alpha1",
+      "kind": "Environment",
+      "metadata": {
+        "name": "artists",
```
Comment (Grendel) [inline packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json:10]: Perhaps `prod` or `qa` is a better example?

File: packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        "description": "Everything about artists"
```
Comment (Grendel) [inline packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json:11]: Perhaps `Production environment` or `Quality Assurance environment`?

File: packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      },
+      "spec": {
+        "owner": "artist-relations-team"
```
Comment (Grendel) [inline packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json:14]: Perhaps `prod-ops-team` or `qa-team`?

Comment (sennyeya) [reply packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json]: this was a holdover from the system that I copied this example from 😓 . Not sure `Environment` should have an owner,  and it's not in the initial spec. Removing for now.

File: packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema",
+  "$id": "EnvironmentV1alpha1",
+  "description": "An Environment represents a single static location that software is deployed to",
+  "examples": [
+    {
+      "apiVersion": "backstage.io/v1alpha1",
+      "kind": "Environment",
+      "metadata": {
+        "name": "qa",
```
Comment (Grendel) [inline packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json:10]: `prod`? to match your decription?

**Comment by jcarres-mdsol** [comment]
Today I was thinking about this and found the mental model that works for me.

Component:  Userservice, Roleservice , libraries, etc.
Environment: Production, Staging, etc.
Installation: Userservice/Production, Userservice/Staging, Roleservice/Staging, etc.

File: packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json (hunk 7 above)
Comment (sennyeya) [reply packages/catalog-model/src/schema/kinds/Environment.v1alpha1.schema.json]: 😅 , yup, updated

**Comment by sennyeya** [comment]
@jcarres-mdsol Thoughts on Installation just being the relation between Component and Environment? As it would be the join between the two, it might make sense to codify it on the relationship instead of a separate deeply coupled entity. What metadata would you want on Installation?

**Comment by sennyeya** [comment]
@Grendel Thanks for chiming in, was hoping you'd take a look! Agreed, I think the environment level views would be really powerful.

I think I'm going to rework this PR to do the metadata on relations approach and see where it ends up. Might end up with too much flexibility, but I think it is a better override approach overall.

**Comment by AndrewAltimit** [comment]
I'm also interested in an environment kind and how I could apply it within my org. I'm thinking it would be possible for this environment kind to represent all 3 environment types (SDLC, ephemeral, personal) listed on the original PR linked in the description. My org uses standardized environments (dev, prod) and environments within those standardized ones (dev-main, dev-lab) so I'm considering how those environments could be represented as entities and tied to their proper components.

Any opinions on whether this PR would enable the following approach? I'm hoping to track prod vs nonprod in either spec.lifecycle or with annotations.

```yaml
apiVersion: backstage.io/v1alpha1
kind: Environment
metadata:
  name: dev-main
  description: Main environment in dev.
spec:
  lifecycle: dev
```

```yaml
apiVersion: backstage.io/v1alpha1
kind: Environment
metadata:
  name: dev-lab
  description: Lab environment in dev.
spec:
  lifecycle: dev
```

```yaml
apiVersion: backstage.io/v1alpha1
kind: Environment
metadata:
  name: prod
  description: Production SDLC environment.
spec:
  lifecycle: production
```

**Comment by sennyeya** [comment]
@AndrewAltimit Hmm, what do you think of a new `partOf` relation on Environment for any sub-environments/ephemeral things? I'm a bit wary of using `spec.lifecycle` or an annotation to track something that could be captured automatically with a relation. Using that relation would mean that you'd need to create a set of environments for SDLC and then model sub-environments as children. As is, this PR wouldn't enable that, but it would be straightforward to add either in this PR or locally in your instance.

**Comment by sennyeya** [comment]
@jcarres-mdsol Not quite following, but if you mean how do we start using metadata on relations -- I'm not sure, still thinking about how to implement that and what that would look like. Ideally, it would mirror existing functions around the catalog just storing pointers to other data and not storing something directly like cost information.

**Comment by jcarres-mdsol** [comment]
I'm saying it is not possible to add metadata in relationships.
So it is not about having a relationship between UserService and Production
but about having a UserServiceProduction entity which is the one that will have annotations

**Comment by mhitkax** [comment]
Late to the party. For our purposes I implemented “environments” and “installations” with custom kinds (using namespaces to avoid name clashes) and custom relationships, similar to what @jcarres-mdsol is describing.

Components and Systems have relationships to the respective Installations, making it easy to show in the System and Component entity pages an overview of their installations with deployed versions and whatnot. The Installation entity pages show the information specific for the given “installation” - status, version, deployment date, I can use existing plugins to show metrics etc. in these pages since they are entities like any other. Creating a table of installations which can filter or sort by status or deployment was a piece of cake thanks to using the catalog (looking forward to paging though). And installations being independent entities, they are listed also in case the relevant Components and/or Systems are not in the catalog yet. I’ll be happy to go more detail if anybody is interested.

@sennyeya Is it worth trying to fit all of that metdata in relationships for this usecase? Why is it better to model it that way? I’d need a rich API around relationships if I were to go that route, plugins will need extra work to support it, and I guess it still won’t work as well in the case if the Component is not in the catalog.

I agree that it’s worth having Environments (it’s very useful for us), but I think that it’s more practical to model it with two parts: Component - Installation - Environment.

BTW I’d still love to be able to add metadata to relationships, but for other usecases (I’ll comment in the BEP).

**Comment by sennyeya** [comment]
@mhitkax Appreciate you chiming in! I'm going to respond to both of your comments here as they're both related. 

I think the `deploysTo` naming is distracting from this work. This has been alluded to above. My goal with this PR wasn't to support modeling deployments but rather provide a way to map components to environments. I think deployments could be layered on top but IMO aren't necessary in the core definition. Every company does them a little differently and I think it's more trouble than it's worth. 

A lot of my current thinking for the direction of this PR is similar to what freben posted [above](https://github.com/backstage/backstage/pull/21560#issuecomment-1826773724),
> What seems to be the bigger need, is to represent aspects of your entity info that actually are different across envs. Like, having a metadata.links entry to the appropriate monitoring page, that's different for each env, or defining different lifecycle value for each. That, to me, starts to sound like an entirely different, larger, problem space that needs even more foundational changes.

That's why I'm working on #22358 and #22130, deployments feels like a harder problem space to get right and environments solve more than just deployment use cases.

**Comment by efenner-cambia** [comment]
Not stale

**Comment by Grendel** [comment]
> @sennyeya hey! Naming is hard. To be clear, I am not interested in deployments as the “act of deploying” itself, but in representing in Backstage what is running / has been deployed to which Environment, which is what you are interested in as well. In my understanding that’s also what @jcarres-mdsol means by an Installation.
> 
> In the comment https://github.com/backstage/backstage/pull/22358/files#r1470326499 I used name ComponentDeployment for the same thing - I guess bad name, again understand that as Installation, ComponentInstance, ComponentInEnvironment…
> 
> My point here is that I see Environments as useful (for overview), I see metadata on relations as useful (for simpler cases), but using metadata on relations to capture the relationship between a Component and an Environment does not seem to me like the best fit.
> 
> Using an entity to represent that thing (AWS lambda or AWS ECS service, Kubernetes Deployment, …) that was created in the environment makes more sense to me (they have an identity, useful metadata, they even might have an owner) and allows to use existing plugins without any need to make them environment-aware since I can add the necessary metadata when generating the catalog info files (or enrich them in the processor). The existing catalog API allows me already now to:
> 
> * track ownership of such component instances across all environments (and show a summary in the group/user pages)
> * show overview of component versions in the Environment page (@Grendel)
> * show overview of component versions across all environments (shown on a tab of Component page)
> * sort component instances by deployment date (@Grendel)
> * filter component instances by region or cluster (as mentioned by @freben) or any arbitrary label
> 
> I’m trying to understand what you see as the benefits of trying to represent those actual component instances as relationships. As I see it, that approach requires a complex metadata structure (again, I’d love to have metadata on relationships, just not for this), potentially rich API around relationships (if you are interested in searching component instances by metadata cross-environment) and requires all relevant plugins to be aware of environment specific metadata. Also, retrieving the Environment entity would also retrieve the full metadata of all component instances in it (since that metadata would be part of the relationships), right?

I think I have come around to this way of thinking... seems like extending relationships is:

A) A big/fundamental extension to Backstages current catalog model which presumably has impacts all over the place
B) Obscures the _thing_ that is an instance of a component deployed to a specific environment

This statement kind of nails it for me (thanks @mhitkax ): "they have an identity, useful metadata, they even might have an owner".

So if I had to restate my original proposal - I am looking for both Environment and DeployedComponents(*) as kinds in the catalog:

[component] - hasDeployedComponent -> [deployedComponent]
[deployedComponent] - instanceOf -> [component]
[deployedComponent] - deployedTo -> [environment]
[environment] - hasDeployedComponent -> [deployedComponent]

I can see [deployedComponent] being populated / updated based on GitHub deployment events... which is tied to a specific tag/commit hash, deployment date, status, etc etc...

(*) I could work with Installation if that term works better for folks

**Comment by aramissennyeydd** [comment]
(same @sennyeya, new account) @Grendel @mhitkax Appreciate your responses here, wanted to jump in and share some thoughts on my side. One of the confusing pieces with this work is that I didn't have a super clear idea of what I was getting into when I started it. My goal has been to add just Environments. I wanted to add Environments because I think the frontend benefit of that is strong for plugins that only support a single annotation and leave developers forced to put the production value with no ability for lower environments.

Deployments got attached somewhere along the way and I think it's a pretty separate idea. To me, that's a much larger can of worms than adding a new  environment kind. Don't get me wrong, it is a conversation that I'm interested in engaging in and getting the platform to adopt, but I don't think this work is the right avenue for that discussion. If one of you wants to open a separate RFC with this PR in mind, please feel free :). I'd be happy to share my thoughts there. I think that work can happen in parallel or sequenced with this.

One of the things that I think differentiates my thinking of Environment with the idea of a Deployment or an Installation, is that the config I'm looking to  overwrite is arguably consistently human-supplied or maintained and generally unlikely to change frequently. The mapping from Component to Environment is direct, ie "Component x has these overrides in these environments". Deployments or Installations are generally autogenerated or sourced from external systems and may change frequently or infrequently depending on your company's deployment cycle. Your company may have specific deployment customizations that require custom properties, annotations, etc. The mapping from Component to Environment uses an intermediate entity, so "Component x has Installation y that is deployed in dev, test and production environments" or even "Component x has 2 Installation y that is deployed in dev and Installation z that is deployed in test and production".

It is possible for these 2 ideas, config-overrides and Deployments, to live symbiotically. One is developers defining specific metadata overrides for their component in a specific environment, the other is developers using Backstage for deployment tasks, Ops tasks or dashboards. 

I think the value add of this PR alone is solid. There are a fair number of plugins like Kubernetes, Lambda, S3, CloudCost, and a few others that would benefit from this kind of implementation.

**Comment by efenner-cambia** [comment]
I agree with @aramissennyeydd that the environment concept is useful on its own with direct relations to components, and that value can come straight from this PR.

I don't see the value add of keeping track of actual deployments as entities within the Backstage catalog. That information is so dynamic that it seems more appropriate to pull it in from the source of truth in real time.

**Comment by efenner-cambia** [comment]
not stale

**Comment by efenner-cambia** [comment]
Bump

**Comment by mcblum** [comment]
Just wanted to say thank you for your work on this and we're excited for it to be merged! Environments were the last thing remaining on the checklist before implementing backstage.

**Comment by aramissennyeydd** [comment]
@efenner-cambia 😞 this is a bit stalled out, I chatted with the catalog folks a few SIGs back and there isn't capacity to review the metadata on relations work that I'd like this PR to leverage. I have to open an RFC for it to be prioritized in the future but haven't found the time. I can separate out the entity itself, but I'm not sure it's a great value add and could sit as a community plugin instead.

**Comment by gioppoluca** [comment]
> > @sennyeya hey! Naming is hard. To be clear, I am not interested in deployments as the “act of deploying” itself, but in representing in Backstage what is running / has been deployed to which Environment, which is what you are interested in as well. In my understanding that’s also what @jcarres-mdsol means by an Installation.
> > In the comment https://github.com/backstage/backstage/pull/22358/files#r1470326499 I used name ComponentDeployment for the same thing - I guess bad name, again understand that as Installation, ComponentInstance, ComponentInEnvironment…
> > My point here is that I see Environments as useful (for overview), I see metadata on relations as useful (for simpler cases), but using metadata on relations to capture the relationship between a Component and an Environment does not seem to me like the best fit.
> > Using an entity to represent that thing (AWS lambda or AWS ECS service, Kubernetes Deployment, …) that was created in the environment makes more sense to me (they have an identity, useful metadata, they even might have an owner) and allows to use existing plugins without any need to make them environment-aware since I can add the necessary metadata when generating the catalog info files (or enrich them in the processor). The existing catalog API allows me already now to:
> > 
> > * track ownership of such component instances across all environments (and show a summary in the group/user pages)
> > * show overview of component versions in the Environment page (@Grendel)
> > * show overview of component versions across all environments (shown on a tab of Component page)
> > * sort component instances by deployment date (@Grendel)
> > * filter component instances by region or cluster (as mentioned by @freben) or any arbitrary label
> > 
> > I’m trying to understand what you see as the benefits of trying to represent those actual component instances as relationships. As I see it, that approach requires a complex metadata structure (again, I’d love to have metadata on relationships, just not for this), potentially rich API around relationships (if you are interested in searching component instances by metadata cross-environment) and requires all relevant plugins to be aware of environment specific metadata. Also, retrieving the Environment entity would also retrieve the full metadata of all component instances in it (since that metadata would be part of the relationships), right?
> 
> I think I have come around to this way of thinking... seems like extending relationships is:
> 
> A) A big/fundamental extension to Backstages current catalog model which presumably has impacts all over the place B) Obscures the _thing_ that is an instance of a component deployed to a specific environment
> 
> This statement kind of nails it for me (thanks @mhitkax ): "they have an identity, useful metadata, they even might have an owner".
> 
> So if I had to restate my original proposal - I am looking for both Environment and DeployedComponents(*) as kinds in the catalog:
> 
> [component] - hasDeployedComponent -> [deployedComponent] [deployedComponent] - instanceOf -> [component] [deployedComponent] - deployedTo -> [environment] [environment] - hasDeployedComponent -> [deployedComponent]
> 
> I can see [deployedComponent] being populated / updated based on GitHub deployment events... which is tied to a specific tag/commit hash, deployment date, status, etc etc...
> 
> (*) I could work with Installation if that term works better for folks

I like this approach: the concept of deployedComponent is going in the right direction.

**Comment by aramissennyeydd** [comment]
Gonna close this for now, I think there's still a few missing pieces from the catalog side and the idea of an environment kind could be extracted into a plugin.

**Comment by homeaidepi** [comment]
Closed?   No environment kind and no way in the UI to bundle versions across an API.  What in the world has this project become?   Its turned into a Document Store like Sharepoint rather then a living System Overview with abilities to view interconnections across boundaries. Well versions and environments are that and based on that "view" that "kind" we need the ability to organize and categorize by these kinds.  Please Add Version and Environment, what is the hold up?

**Comment by nia-potato** [comment]
hi @aramissennyeydd, I also dont think this should be closed. 

if this is closed, what is the missing pieces or PR/Issues that are missing so we can get this move forward? 


For backstage to be a system catalog although it contains the class (root) level info of every service, but without containing the object level resources in a relation with the environments has very little meaning to the developers since it still doesnt solve the problem where each classes object is being used and in what environment is being used. they still dont know where things are even if you just give them a overview class component.

**Comment by jcarres-mdsol** [comment]
I agree, some codebases only become a service when deployed into an environment. 
Maybe you have a "User service" which is the same in all your environments. Then you may think you do not need this.
But some codebases become "Enduser input service" in one environment and "backend operation" in another. So it is really messy to model stuff without environments to help.

**Comment by aramissennyeydd** [comment]
@jcarres-mdsol @nia-potato Please continue conversation in https://github.com/backstage/backstage/issues/16389. Environment seems like a simple concept but as we saw in the thread above is pretty company specific. I'd recommend starting with adding this type internally and then considering what OSS work is needed to support additional use cases where necessary.

**Pull request closed without merging**